ext/rotoscope/rotoscope.c

```diff
@@ -133,30 +133,24 @@ static rs_tracepoint_t extract_full_tracevals(rb_trace_arg_t *trace_arg,
 
 static bool in_fork(Rotoscope *config) { return config->pid != getpid(); }
 
-static bool tracecmp(rs_tracepoint_t *a, rs_tracepoint_t *b) {
-  return (!rb_str_cmp(a->method_name, b->method_name) &&
-          !rb_str_cmp(a->entity, b->entity) &&
-          a->method_level == b->method_level);
-}
-
-static void log_raw_trace(FILE *stream, rs_tracepoint_t trace) {
+static void log_trace_event(FILE *stream, rs_tracepoint_t *trace) {
   fprintf(stream, RS_CSV_FORMAT "\n", RS_CSV_VALUES(trace));
 }
 
-unsigned char output_buffer[500];
-static void log_stack_frame(FILE *stream, rs_stack_t *stack,
-                            rs_strmemo_t **call_memo, rs_tracepoint_t trace,
-                            rb_event_flag_t event) {
-  if (event & EVENT_CALL) {
-    rs_stack_frame_t frame = rs_stack_push(stack, trace);
-    sprintf((char *)output_buffer, RS_FLATTENED_CSV_FORMAT "\n",
-            RS_FLATTENED_CSV_VALUES(frame));
-
-    if (rs_strmemo_uniq(call_memo, output_buffer)) {
-      fputs((char *)output_buffer, stream);
-    }
-  } else if (event & EVENT_RETURN) {
-    if (tracecmp(&trace, &rs_stack_peek(stack)->tp)) rs_stack_pop(stack);
+unsigned char output_buffer[LOG_BUFFER_SIZE];
+static void log_trace_event_with_caller(FILE *stream,
+                                        rs_stack_frame_t *stack_frame,
+                                        rs_strmemo_t **call_memo) {
+  rs_stack_frame_t *caller = stack_frame->caller;
+  while (caller->blacklisted) {
+    caller = caller->caller;
```

Comment:
Any chance of walking all the way back up the stack to where `caller->caller` is null?

Reply:
The stack inserts a root node in `insert_root_node` and this PR makes it not blacklisted so we wouldn't have to worry about dereferencing NULL.